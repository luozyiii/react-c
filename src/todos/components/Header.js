import React, { Component } from 'react';

export default class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleInputKeyUp = this.handleInputKeyUp.bind(this);
  }

  handleInputChange(e) {
    this.setState({
      value: e.target.value,
    });
  }

  handleInputKeyUp(e) {
    const { value } = this.state;
    if (e.keyCode === 13 && value) {
      this.props.addUndoItem(value);
      this.setState({
        value: '',
      });
    }
  }

  render() {
    const { value } = this.state;
    return (
      <div className="header">
        <div className="header-content">
          Todos <input placeholder="add todo" className="header-input" value={value} onChange={this.handleInputChange} onKeyUp={this.handleInputKeyUp} data-test="input" />
        </div>
      </div>
    );
  }
}
