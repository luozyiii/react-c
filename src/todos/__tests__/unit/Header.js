import React from 'react';
import { shallow } from 'enzyme';
import { expect } from '@jest/globals';
import { findTestWrapper } from '../../../utils/testUtils';
import Header from '../../components/Header';

it('Header 渲染样式正常', () => {
  const wrapper = shallow(<Header />);
  expect(wrapper).toMatchSnapshot();
});

it('Header 组件包含一个 input 框', () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, 'input');
  expect(inputElem.length).toBe(1);
});

it('Header 组件 input 框内容，初始化应该为空', () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, 'input');
  expect(inputElem.prop('value')).toEqual('');
});

it('Header 组件 input 框内容，当用户输入时，会跟随变化', () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, 'input');
  const userInput = '今天要学习 Jest';
  inputElem.simulate('change', {
    target: { value: userInput },
  });
  // expect(wrapper.state('value')).toEqual(userInput); // state 只能在class 组件中使用
  const newInputElem = wrapper.find('[data-test="input"]');
  expect(newInputElem.prop('value')).toEqual(userInput);
});

it('Header 组件 input 框输入回车时，如果 input 无内容，无操作', () => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const inputElem = findTestWrapper(wrapper, 'input');
  // setState 只能在class 组件中使用
  wrapper.setState({
    value: '',
  });
  inputElem.simulate('keyUp', {
    keyCode: 13,
  });
  expect(fn).not.toHaveBeenCalled();
});

it('Header 组件 input 框输入回车时，如果 input 有内容，函数应该被调用', () => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const inputElem = findTestWrapper(wrapper, 'input');
  const userInput = '学习 React';
  // setState 只能在class 组件中使用
  wrapper.setState({
    value: userInput,
  });
  inputElem.simulate('keyUp', {
    keyCode: 13,
  });
  expect(fn).toHaveBeenCalled();
  expect(fn).toHaveBeenLastCalledWith(userInput);
});

it('Header 组件 input 框输入回车时，如果 input 无内容，最后应该清除掉', () => {
  const fn = jest.fn();
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, 'input');
  const userInput = '';
  // setState 只能在class 组件中使用
  wrapper.setState({
    value: userInput,
  });
  inputElem.simulate('keyUp', {
    keyCode: 13,
  });
  const newInputElem = wrapper.find('[data-test="input"]');
  expect(newInputElem.prop('value')).toBe('');
});
