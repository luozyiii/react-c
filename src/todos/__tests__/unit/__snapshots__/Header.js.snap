// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 组件 渲染样式正常 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    Todos 
    <input
      className="header-input"
      data-test="header-input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="add todo"
      value=""
    />
  </div>
</div>
`;
